dotnet/src/SemanticKernel.Skills/Skills.Memory.CosmosDB/CosmosMemoryStore.cs

```diff
@@ -0,0 +1,207 @@
+            }
+        }
+
+        embeddings.SortByScore();
+
+        return embeddings.Select(x => (x.Value, x.Score.Value)).ToAsyncEnumerable();
+    }
+
+    protected async IAsyncEnumerable<DataEntry<IEmbeddingWithMetadata<TEmbedding>>> TryGetCollectionAsync(string collectionName, [EnumeratorCancellation] CancellationToken cancel = default)
+    {
+        var container = this._client.GetContainer(this._databaseName, this._containerName);
+        var query = new QueryDefinition($"SELECT * FROM c WHERE c.collectionId = '{collectionName}'");
+        var iterator = container.GetItemQueryIterator<CosmosMemoryRecord>(query);
+
+        var items = await iterator.ReadNextAsync(cancel).ConfigureAwait(false);
+
+        foreach (var item in items)
+        {
+            var embeddingHost = JsonConvert.DeserializeAnonymousType(
+                item.EmbeddingString,
+                new { Embedding = new { vector = new List<float>() } });
+
+            var rec = MemoryRecord.FromJson(
+                item.MetadataString,
+                new Embedding<float>(embeddingHost.Embedding.vector));
+
+            yield return DataEntry.Create<IEmbeddingWithMetadata<TEmbedding>>(
+                rec.Metadata.Id,
+                (IEmbeddingWithMetadata<TEmbedding>)rec,
+                item.Timestamp);
+        }
+    }
+
+    public async Task<DataEntry<IEmbeddingWithMetadata<TEmbedding>>> PutAsync(string collection, DataEntry<IEmbeddingWithMetadata<TEmbedding>> data, CancellationToken cancel = default)
+    {
+        var entity = new CosmosMemoryRecord
+        {
+            CollectionId = collection,
+            Id = this._toCosmosFriendlyId(data.Key),
+            Timestamp = data.Timestamp,
+            EmbeddingString = data.ValueString!,
+            MetadataString = data.Value!.GetSerializedMetadata()
+        };
+
+        var container = this._client.GetContainer(this._databaseName, this._containerName);
+
+        await container.CreateItemAsync(entity, cancellationToken: cancel, requestOptions: new ItemRequestOptions()
+        {
+            EnableContentResponseOnWrite = false
+        });
+
+        return data;
+    }
+
+    public Task RemoveAsync(string collection, string key, CancellationToken cancel = default)
+    {
+        var container = this._client.GetContainer(this._databaseName, this._containerName);
+
+        return container.DeleteItemAsync<CosmosMemoryRecord>(
+            this._toCosmosFriendlyId(key),
+            new Microsoft.Azure.Cosmos.PartitionKey(collection),
+            cancellationToken: cancel);
+    }
+
+    private string _toCosmosFriendlyId(string id)
+    {
+        return $"{id.Trim().Replace(' ', '-').Replace('/', '_').Replace('\\', '_').Replace('?', '_').Replace('#', '_').ToUpperInvariant()}";
+    }
+
+    protected virtual void Dispose(bool disposing)
+    {
+        if (!this._disposedValue)
+        {
+            if (disposing)
+            {
+                this._client.Dispose();
```

Comment:
The client is created outside of this class and, in theory, can be shared between a few instances of the CosmosMemoryStore class and if one of the instances is disposed the other ones won't be able to use the client anymore. I think, the client should be disposed outside of this class, somewhere up the stack in a place where there's enough information to make such a decision, host app?

Reply:
You're right, good call!